cpp/src/parquet/column_reader_test.cc

```diff
@@ -573,5 +573,464 @@ TEST_F(TestPrimitiveReader, TestNonDictionaryEncodedPagesWithExposeEncoding) {
+
+  ASSERT_EQ(records_read, 2);
+  ASSERT_EQ(record_reader->values_written(), 2);
+  ASSERT_EQ(record_reader->levels_written(), 0);
+  ASSERT_EQ(record_reader->levels_position(), 0);
+
+  const auto read_values =
+      reinterpret_cast<const int32_t*>(record_reader->values());
+  std::vector<int32_t> read_vals(read_values,
+                                 read_values + record_reader->values_written());
+
+  ASSERT_TRUE(vector_equal(read_vals, {30, 30}));
+}
+
+// Skip an optional field. Intentionally included some null values.
+TEST(RecordReaderTest, SkipOptional) {
+  internal::LevelInfo level_info;
+  level_info.def_level = 1;
+  level_info.rep_level = 0;
+
+  NodePtr type = schema::Int32("b", Repetition::OPTIONAL);
+  const ColumnDescriptor descr(type, level_info.def_level,
+                               level_info.rep_level);
+
+  // Records look like {null, 20, 20, 20, null, 30, 30}
+  std::vector<std::shared_ptr<Page>> pages;
+  std::vector<int32_t> values = {20, 20, 20, 30, 30};
+  std::vector<int16_t> def_levels = {0, 1, 1, 1, 0, 1, 1};
+
+  std::unique_ptr<PageReader> pager;
+  std::shared_ptr<DataPageV1> page = MakeDataPage<Int32Type>(
+      &descr, values, /*num_values=*/values.size(), Encoding::PLAIN,
+      /*indices=*/{},
+      /*indices_size=*/0, def_levels, level_info.def_level,
+      /*rep_levels=*/{}, level_info.rep_level);
+  pages.push_back(std::move(page));
+  pager.reset(new test::MockPageReader(pages));
+
+  std::shared_ptr<internal::RecordReader> record_reader =
+      internal::RecordReader::Make(&descr, level_info);
+  record_reader->SetPageReader(std::move(pager));
+
+  // This also tests when we start with a Skip.
+  int64_t records_skipped = record_reader->SkipRecords(/*num_records=*/2);
+
+  ASSERT_EQ(records_skipped, 2);
+  ASSERT_EQ(record_reader->values_written(), 0);
+  // Since we started with skipping, there was nothing in the buffer to consume
+  // for skipping.
+  ASSERT_EQ(record_reader->levels_written(), 0);
+  ASSERT_EQ(record_reader->levels_position(), 0);
+
+  int64_t records_read = record_reader->ReadRecords(/*num_records=*/3);
+
+  ASSERT_EQ(records_read, 3);
+  // One of these values is null. values_written() includes null values.
+  ASSERT_EQ(record_reader->values_written(), 3);
+  // When we read, we buffer some levels. Since there are only a few levels
+  // in our whole column, all of them are read.
+  // We had skipped 2 of the levels above. So there is only 5 left in total to
+  // read, and we read 3 of them here.
+  ASSERT_EQ(record_reader->levels_written(), 5);
+  ASSERT_EQ(record_reader->levels_position(), 3);
+
+  const auto read_values =
+      reinterpret_cast<const int32_t*>(record_reader->values());
+  std::vector<int32_t> read_vals(read_values,
+                                 read_values + record_reader->values_written() -
+                                     record_reader->null_count());
+
+  ASSERT_TRUE(vector_equal(read_vals, {20, 20}));
```

Comment:
GMock should be available in arrow, so it would be clearer to use EXPECT_THAT(read_values, ElementsAre(20, 20)); here.

Reply:
Done.